crates/test-programs/wasi-tests/src/bin/dir_fd_op_failures.rs

```diff
@@ -5,6 +5,24 @@ unsafe fn test_fd_dir_ops(dir_fd: wasi::Fd) {
     let stat = wasi::fd_filestat_get(dir_fd).expect("failed to fdstat");
     assert_eq!(stat.filetype, wasi::FILETYPE_DIRECTORY);
 
+    let (pr_fd, pr_name_len) = (3..)
+        .map_while(|fd| wasi::fd_prestat_get(fd).ok().map(|stat| (fd, stat)))
+        .find_map(|(fd, wasi::Prestat { tag, u })| {
+            (tag == wasi::PREOPENTYPE_DIR.raw()).then_some((fd, u.dir.pr_name_len))
+        })
+        .expect("failed to find preopen directory");
+
+    let r = wasi::fd_prestat_dir_name(pr_fd, [].as_mut_ptr(), 0);
+    assert_eq!(r, Err(wasi::ERRNO_NAMETOOLONG));
+
+    // Test that passing a larger than necessary buffer works correctly
+    let r = wasi::fd_prestat_dir_name(
+        pr_fd,
+        vec![0; pr_name_len + 1].as_mut_ptr(),
```

Comment:
I believe that this as-written is ok, but I'd recommend avoiding this pattern in the future as it's so easy to get wrong. This only works if the `vec![...]` allocate here lives longer than the function call which I think rustc's temporary rules satisfy here, but I at least personally think it'd be more clear to move the vec above this statement to ensure it lives across the statement.

Reply:
Thanks! Yes, good point, I would guess that `miri` should complain about this or at least I remember seeing it fail on a similar pattern in a different project before.